### PR TITLE
Remove dependency on `base` and `stdio`

### PR DESCRIPTION
**Pull request by phongulus**
## Description of the task

Removed and replaced all uses of the Base and Stdio libraries.

## How to test

Should compile without issues and the tests should pass.

```
make test
```

### CODE REVIEW

**Comment by tysg** [comment]
You mentioned 
```
 "ocaml" {>= "4.08.0"}
```
is too strict. Should we relax this then? What's the minimal required version to compile this library?

File: lib/rule.ml [hunk 1]
```diff
@@ -55,38 +56,42 @@ module Prefix = struct
...
       evaluated before the allow list. *)
   let match_rules filename ~rules =
-    let compare a b = Int.compare (snd a) (snd b) in
-    let is_prefix prefix = String.is_prefix filename ~prefix in
+    let max_elt t =
+      let compare a b = Int.compare (snd a) (snd b) in
+      match t with
+      | [] -> None
+      | v :: vs -> Some (List.fold_left (fun a b -> if compare b a > 0 then b else a) v vs) in
+    let is_prefix prefix = Stre.starts_with filename prefix in
```
Comment (phongulus) [inline lib/rule.ml:64]: One thing I noticed while doing this is that in `monorobot.opam` the minimum required version for ocaml is `>=4.08.0`. There are some API changes in later versions that would make life easier in some places, like `String.starts_with` (here) and `String.fold_left` (`lib/mrkdwn.ml` line 11). Right now they're replaced with equivalents from `Extlib` and `Devkit`.  Should we bump the minimum version up to 4.14? I think that's the compiler installed by default when running `opam switch create ./` anyway.

**Comment by phongulus** [comment]
> You mentioned
> 
> ```
>  "ocaml" {>= "4.08.0"}
> ```
> 
> is too strict. Should we relax this then? What's the minimal required version to compile this library?

I think `4.14.0` is the default opam installs. I tried to see if it compiles with `4.08.0` but I couldn't get the compiler down to that version (M1 Mac only supported from `4.12` - it does compile there though).

File: lib/api_local.ml [hunk 2]
```diff
@@ -65,7 +66,8 @@ module Slack : Api.Slack = struct
   include Slack_base
 
   let lookup_user ?cache:_ ~ctx:_ ~(cfg : Config_t.config) ~email () =
-    let email = List.Assoc.find cfg.user_mappings ~equal:String.equal email |> Option.value ~default:email in
+    let signature_email_opt = List.find_opt (fun (k, _) -> String.equal k email) cfg.user_mappings in
```
Comment (Khady) [inline lib/api_local.ml:69]: List.assoc_opt

File: lib/common.ml [hunk 3]
```diff
@@ -1,23 +1,27 @@
...
 module StringMap = struct
-  type 'a t = 'a Map.M(String).t
+  module Map = Map.Make(String)
+  include Map
+  type 'a t = 'a Map.t
 
-  let empty : 'a t = Map.empty (module String)
-  let to_list (l : 'a t) : (string * 'a) list = Map.to_alist l
-  let of_list (m : (string * 'a) list) : 'a t = Map.of_alist_exn (module String) m
+  let empty : 'a t = Map.empty
```
Comment (Khady) [inline lib/common.ml:8]: aren't you re-defining stuff that is already included in `Map`?

File: lib/api_remote.ml [hunk 4]
```diff
@@ -1,27 +1,26 @@
-open Base
 open Printf
 open Devkit
 open Common
 
 module Github : Api.Github = struct
   let commits_url ~(repo : Github_t.repository) ~sha =
-    String.substr_replace_first ~pattern:"{/sha}" ~with_:("/" ^ sha) repo.commits_url
+    snd @@ ExtLib.String.replace ~sub:"{/sha}" ~by:("/" ^ sha) ~str:repo.commits_url
```
Comment (tysg) [inline lib/api_remote.ml:7]: Using `snd` can cause some confusion to the reader. Consider:
```
let sub = "{/sha}" in
let by = "/" ^ sha in
let _, output = ExtLib.String.replace ~sub ~by ~str:repo.commits_url in
str
```

File: lib/common.ml [hunk 5]
```diff
@@ -53,14 +56,24 @@ let http_request ?headers ?body meth path =
...
 
 let sign_string_sha256 ~key ~basestring =
   Cstruct.of_string basestring |> Nocrypto.Hash.SHA256.hmac ~key:(Cstruct.of_string key) |> Hex.of_cstruct |> Hex.show
+
+let dedup_and_sort ~compare l =
+  List.fold_right
+    (fun s (last, l) ->
+      match last with
+      | None -> (Some s, s :: l)
+      | Some last -> if compare s last == 0 then (Some last, l) else (Some s, s :: l))
```
Comment (Khady) [inline lib/common.ml:76]:
```suggestion
      | Some last -> if compare s last = 0 then (Some last, l) else (Some s, s :: l))
```

File: lib/api_remote.ml [hunk 6]
```diff
@@ -36,7 +35,7 @@ module Github : Api.Github = struct
       | "base64" ->
         begin
           try
-            response.content |> String.split_lines |> String.concat |> decode_string_pad |> Config_j.config_of_string
+            response.content |> String.split_on_char '\n' |> String.concat "" |> decode_string_pad |> Config_j.config_of_string
```
Comment (tysg) [inline lib/api_remote.ml:38]: `String.split_on_char '\n' |> String.concat ""` this is essentially removing all `\n` in the string, but this implementation allocates 2x the size of the original string. Consider using regex, e.g. `Re2.replace`.

File: lib/common.ml (hunk 5 above)
Comment (Khady) [inline lib/common.ml:76]:
```suggestion
      | Some last when compare s last = 0 -> (Some last, l)
      | _ -> (Some s, s :: l)
```

File: lib/github.ml [hunk 7]
```diff
@@ -75,9 +74,9 @@ let validate_signature ?signing_key ~headers body =
   match signing_key with
   | None -> Ok ()
   | Some secret ->
-  match List.Assoc.find headers "x-hub-signature" ~equal:String.equal with
+  match List.find_opt (fun (k, _) -> String.equal k "x-hub-signature") headers with
```
Comment (Khady) [inline lib/github.ml:77]: List.assoc_opt

File: lib/github.ml [hunk 8]
```diff
@@ -103,12 +102,12 @@ let parse_exn headers body =
     | Pending -> "pending"
     | Error -> "error"
   in
-  let print_opt f v = Option.value_map ~f ~default:"none" v in
+  let print_opt f v = Option.map_default f "none" v in
   let print_comment_preview = Stre.shorten ~escape:true 40 in
-  let print_commit_hash s = String.prefix s 8 in
-  match List.Assoc.find headers "x-github-event" ~equal:String.equal with
+  let print_commit_hash s = String.sub s 0 (min 8 @@ String.length s) in
+  match List.find_opt (fun (k, _) -> String.equal k "x-github-event") headers with
```
Comment (Khady) [inline lib/github.ml:108]: List.assoc_opt

File: lib/api_remote.ml [hunk 9]
```diff
@@ -121,7 +120,8 @@ module Slack : Api.Slack = struct
 
   let lookup_user' ~(ctx : Context.t) ~(cfg : Config_t.config) ~email () =
     (* Check if config holds the Github to Slack email mapping  *)
-    let email = List.Assoc.find cfg.user_mappings ~equal:String.equal email |> Option.value ~default:email in
+    let signature_email_opt = List.find_opt (fun (k, _) -> String.equal k email) cfg.user_mappings in
+    let email = Option.map_default snd email signature_email_opt in
```
Comment (tysg) [inline lib/api_remote.ml:124]: OCaml stdlib has `List.Assoc.assoc_opt` too.

File: lib/github.ml [hunk 10]
```diff
@@ -182,15 +181,16 @@ let gh_link_of_string url_str =
   match Uri.host url with
   | None -> None
   | Some host ->
-  match String.chop_prefix path ~prefix:"/" with
+  match (if Stre.starts_with path "/" then Some (Stre.drop_prefix path "/") else None) with
+  (* match String.chop_prefix path ~prefix:"/" with *)
```
Comment (Khady) [inline lib/github.ml:185]:
```suggestion
  match Stre.starts_with path "/" with
  | false -> None
  | true ->
  let path = Stre.drop_prefix path "/"
```

File: lib/github.ml [hunk 11]
```diff
@@ -182,15 +181,16 @@ let gh_link_of_string url_str =
...
   | Some path ->
-    let path = String.chop_suffix_if_exists ~suffix:"/" path |> flip Stre.nsplitc '/' |> List.map ~f:Web.urldecode in
+    let path = Stre.drop_suffix path "/" |> flip Stre.nsplitc '/' |> List.map Web.urldecode in
     let make_repo ~prefix ~owner ~name =
-      let base = String.concat ~sep:"/" (List.rev prefix) in
+      let base = String.concat "/" (List.rev prefix) in
       let scheme = Uri.scheme url in
       let html_base, api_base =
-        if String.is_suffix base ~suffix:"github.com" then gh_com_html_base owner name, gh_com_api_base owner name
+        if Stre.drop_suffix base "github.com" != base then gh_com_html_base owner name, gh_com_api_base owner name
```
Comment (Khady) [inline lib/github.ml:193]:
```suggestion
        if Stre.drop_suffix base "github.com" <> base then gh_com_html_base owner name, gh_com_api_base owner name
```

File: lib/context.ml [hunk 12]
```diff
@@ -69,7 +68,7 @@ let is_pipeline_allowed ctx repo_url ~pipeline =
   | None -> true
   | Some config ->
   match config.status_rules.allowed_pipelines with
-  | Some allowed_pipelines when not @@ List.exists allowed_pipelines ~f:(String.equal pipeline) -> false
+  | Some allowed_pipelines when not @@ List.mem pipeline allowed_pipelines -> false
```
Comment (tysg) [inline lib/context.ml:71]: `List.mem` uses `(=)` to test equality, consider using `List.find_opt` or `List.exists` to be more explicit on which equal function to use.

File: lib/action.ml [hunk 13]
```diff
@@ -140,14 +139,14 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
...
-        let has_same_status_state_as_prev (branch : branch) =
-          match Map.find branch_statuses branch.name with
+        let has_different_status_state_from_prev (branch : branch) =
+          not @@ match StringMap.find_opt branch.name branch_statuses with
           | None -> false
-          | Some state -> Poly.equal state current_status
+          | Some state -> state = current_status
         in
-        let branches = List.filter n.branches ~f:(Fn.non @@ has_same_status_state_as_prev) in
+        let branches = List.filter has_different_status_state_from_prev n.branches in
```
Comment (tysg) [inline lib/action.ml:149]: OCaml has `Fun.negate` so you can avoid slapping a `not` before a match expression.

File: lib/action.ml [hunk 14]
```diff
@@ -376,14 +376,14 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
           | Some id -> Lwt.return_some id
           | None -> fetch_bot_user_id ()
         in
-        Lwt.return (Option.value_map ~default:false ~f:(String.equal event.user) bot_user_id)
+        Lwt.return (Option.map (String.equal event.user) bot_user_id |> Option.default false)
```
Comment (tysg) [inline lib/action.ml:379]: `Option.map_default`

File: src/monorobot.ml [hunk 15]
```diff
@@ -56,7 +55,7 @@ let check_gh_action file json config secrets state =
     )
 
 let check_slack_action file secrets =
-  let data = Stdio.In_channel.read_all file in
+  let data = ExtLib.input_file file in
```
Comment (tysg) [inline src/monorobot.ml:58]: Stdlib can do it too `In_channel.(with_open_bin file input_all)`

**Review by tysg — CHANGES_REQUESTED** [review]

File: lib/github.ml (hunk 11 above)
Comment (phongulus) [reply lib/github.ml:193]: Is there a difference between `!=` and `<>` if we're comparing non-mutables? Or is it a style/convention thing?

Comment (Khady) [reply lib/github.ml:193]:
```
# let s1 = "a" ^ "b";;
val s1 : string = "ab"
# let s2 = "a" ^ "b";;
val s2 : string = "ab"
# s1 = s2;;
- : bool = true
# s1 == s2;;
- : bool = false
# s1 <> s2;;
- : bool = false
# s1 != s2;;
- : bool = true
```

File: lib/common.ml [hunk 16]
```diff
@@ -1,23 +1,27 @@
...
+  let of_list_multi (m : (string * 'a) list) : 'a list t =
+    let update v = function None -> Some [v] | Some vs -> Some (v :: vs) in
+    List.fold_right (fun (k, v) b -> Map.update k (update v) b) m empty
 end
 
 module Stringtbl = struct
   include Hashtbl
+  type 'a t = (string, 'a) Hashtbl.t
 
-  type 'a t = 'a Hashtbl.M(String).t
```
[the remark this answers is not in the document]
Comment (phongulus) [reply lib/common.ml:16]: `Hashtbl.Make` requires defining a hash module for string I believe: https://v2.ocaml.org/api/Hashtbl.HashedType.html

File: lib/api_remote.ml (hunk 4 above)
Comment (phongulus) [reply lib/api_remote.ml]: I feel like OCaml `snd` should be fairly understood? Would it better to put it in the back, like `|> snd` for readability?

File: src/monorobot.ml (hunk 15 above)
Comment (phongulus) [reply src/monorobot.ml:58]: `In_channel` is in OCaml from `4.14` onward, so we can probably use it if we bump up the require version?

File: lib/action.ml [hunk 17]
```diff
@@ -12,38 +11,38 @@ let log = Log.from "action"
 
 module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
   let partition_push (cfg : Config_t.config) n =
-    let default = Option.to_list cfg.prefix_rules.default_channel in
+    let default = Stdlib.Option.to_list cfg.prefix_rules.default_channel in
```
[the remark this answers is not in the document]
Comment (phongulus) [reply lib/action.ml:14]: I think `Extlib` somehow overrides `Stdlib` here, it doesn't compile unless I add this.

File: lib/common.ml [hunk 18]
```diff
@@ -1,23 +1,27 @@
...
-
-  let empty () = Hashtbl.create (module String)
-  let to_list (l : 'a t) : (string * 'a) list = Hashtbl.to_alist l
-  let of_list (m : (string * 'a) list) : 'a t = Hashtbl.of_alist_exn (module String) m
+  let empty () = Hashtbl.create 0
```
[the remark this answers is not in the document]
Comment (phongulus) [reply lib/common.ml]: I'm not sure how big the initial hash table should be - it's being used to configs for different repos?

File: lib/api_remote.ml (hunk 4 above)
Comment (tysg) [reply lib/api_remote.ml]: It is not immediately clear what the first and second element in the tuple represent from the reader's POV. Assigning the pattern-matched value a variable name helps the reader to understand the code better.

File: lib/github.ml [hunk 19]
```diff
@@ -182,15 +181,16 @@ let gh_link_of_string url_str =
...
   | Some host ->
-  match String.chop_prefix path ~prefix:"/" with
-  | None -> None
-  | Some path ->
-    let path = String.chop_suffix_if_exists ~suffix:"/" path |> flip Stre.nsplitc '/' |> List.map ~f:Web.urldecode in
+  match Stre.starts_with path "/" with
+  | false -> None
+  | true ->
+    let path =
+      Stre.drop_suffix (Stre.drop_prefix path "/") "/" |> flip Stre.nsplitc '/' |> List.map Web.urldecode in
```
Comment (Khady) [inline lib/github.ml:188]:
```suggestion
      path |> flip Stre.drop_prefix "/" |> flip Stre.drop_suffix "/" |> flip Stre.nsplitc '/' |> List.map Web.urldecode in
```

File: lib/common.ml (hunk 18 above)
[the remark this answers is not in the document]
Comment (tysg) [reply lib/common.ml]: A small number, e.g. 10, will do. Although interestingly `base` uses 0 ([ref](https://github.com/janestreet/base/blob/1311ac11d54b79698c8d559d1334b0b815757788/src/hashtbl.ml#L55)).

[the remark this answers is not in the document]
Comment (phongulus) [reply lib/common.ml]: If I understand correctly the Stringtbl is used to store per monorepo configs. Unless it's expected that we monitor multiple monorepos.

File: lib/github.ml [hunk 20]
```diff
@@ -182,15 +182,17 @@ let gh_link_of_string url_str =
...
+    let path =
+      Stre.drop_prefix path "/" |> flip Stre.drop_suffix "/" |> flip Stre.nsplitc '/' |> List.map Web.urldecode
+    in
     let make_repo ~prefix ~owner ~name =
-      let base = String.concat ~sep:"/" (List.rev prefix) in
+      let base = String.concat "/" (List.rev prefix) in
       let scheme = Uri.scheme url in
       let html_base, api_base =
-        if String.is_suffix base ~suffix:"github.com" then gh_com_html_base owner name, gh_com_api_base owner name
+        if Stre.drop_suffix base "github.com" <> base then gh_com_html_base owner name, gh_com_api_base owner name
```
Comment (rr0gi) [inline lib/github.ml:195]: this is confusing - String.ends_with would be easier to read

File: lib/mrkdwn.ml [hunk 21]
```diff
@@ -1,14 +1,14 @@
...
 let escape_mrkdwn =
-  String.concat_map ~f:(function
+  let c_map = function
     | '<' -> "&lt;"
     | '>' -> "&gt;"
     | '&' -> "&amp;"
     | c -> String.make 1 c
-    )
+  in
+  ExtLib.String.fold_left (fun s c -> s ^ c_map c) ""
```
Comment (rr0gi) [inline lib/mrkdwn.ml:11]: concat_map can be implemented efficiently, while fold_left with `^` inside cannot. There is `Stre.catmap` in devkit

File: lib/mrkdwn.ml [hunk 22]
```diff
@@ -108,39 +108,47 @@ let rec mrkdwn_of_md md =
...
       loop list_indent tl
     | Ol l ->
       nl_if_needed_above b;
-      List.iteri l ~f:(fun i li ->
-        add_spaces list_indent;
-        Printf.bprintf b "%d. " (i + 1);
-        loop ~is_in_list:true (list_indent + 4) li;
-        nl_if_needed_above b
-      );
+      List.iteri
```
Comment (rr0gi) [inline lib/mrkdwn.ml:123]:
```suggestion
      l |> List.iteri
```

File: lib/slack_message.ml [hunk 23]
```diff
@@ -157,13 +159,13 @@ let populate_commit ?(include_changes = true) repository (api_commit : api_commi
         but looks like slack doesn't provide such functionality
       *)
       try
-        match String.split commit.author.date ~on:'T' with
+        match String.split_on_char 'T' commit.author.date with
```
Comment (rr0gi) [inline lib/slack_message.ml:162]: this looks like `Devkit.Time` time format parsing, should be moved to devkit

**Review by rr0gi — CHANGES_REQUESTED** [review]

File: lib/rule.ml (hunk 1 above)
Comment (tysg) [reply lib/rule.ml:65]: Yes, let's do that. Thanks!

**Review by tysg — APPROVED** [review]
Should be good to merge apart from 1 minor issue. Thanks!

File: lib/common.ml [hunk 24]
```diff
@@ -53,14 +58,14 @@ let http_request ?headers ?body meth path =
...
-  try Ok (Devkit.Files.save_as path (fun oc -> Stdio.Out_channel.fprintf oc "%s" data))
+  try Ok (Devkit.Files.save_as path (fun oc -> Printf.fprintf oc "%s" data))
   with exn -> fmt_error "%s" (Exn.to_string exn)
 
 let longest_common_prefix xs =
   match xs with
   | [] -> ""
   | [ x ] -> x
-  | x :: _ -> String.sub x ~pos:0 ~len:(Stre.common_prefix x (List.sort xs ~compare:String.compare |> List.last_exn))
+  | x :: _ -> String.sub x 0 (Stre.common_prefix x (List.sort String.compare xs |> List.rev |> List.hd))
```
[the remark this answers is not in the document]
Comment (phongulus) [reply lib/common.ml]: There's a problem with the original code's logic here. Got the following
```
utop # longest_common_prefix ["/test/paths/not_here"; "/test/path/here"; "/test/paths/here"];;
- : string = "/test/paths/not_here"
```
I'll fix it in another PR.

**Pull request merged**